src/v/config/config_store.h

```diff
@@ -108,13 +108,26 @@ class config_store {
         }
     }
 
-    void to_json(rapidjson::Writer<rapidjson::StringBuffer>& w) const {
+    /**
+     *
+     * @param filter optional callback for filtering out config properties.
+     *               callback should return false to exclude a property.
+     */
+    void to_json(
+      rapidjson::Writer<rapidjson::StringBuffer>& w,
+      std::optional<std::function<bool(base_property&)>> filter
```

Comment:
nit: `std::function<>` already behaves like an optional if it doesn't have a value set.